agent/dockerclient/sdkclientfactory/sdkclientfactory.go

```diff
@@ -0,0 +1,198 @@
+	// GetDefaultClient returns a versioned client for the default version
+	GetDefaultClient() (sdkclient.Client, error)
+
+	// GetClient returns a client with the specified version or an error
+	// if the client doesn't exist.
+	GetClient(version dockerclient.DockerVersion) (sdkclient.Client, error)
+
+	// FindSupportedAPIVersions returns a slice of agent-supported Docker API
+	// versions. Versions are tested by making calls against the Docker daemon
+	// and may occur either at Factory creation time or lazily upon invocation
+	// of this function. The slice represents the intersection of
+	// agent-supported versions and daemon-supported versions.
+	FindSupportedAPIVersions() []dockerclient.DockerVersion
+
+	// FindKnownAPIVersions returns a slice of Docker API versions that are
+	// known to the Docker daemon. Versions are tested by making calls against
+	// the Docker daemon and may occur either at Factory creation time or
+	// lazily upon invocation of this function. The slice represents the
+	// intersection of the API versions that the agent knows exist (but does
+	// not necessarily fully support) and the versions that result in
+	// successful responses by the Docker daemon.
+	FindKnownAPIVersions() []dockerclient.DockerVersion
+
+	// FindClientAPIVersion returns the client api version
+	FindClientAPIVersion(sdkclient.Client) dockerclient.DockerVersion
+}
+
+type factory struct {
+	endpoint string
+	clients  map[dockerclient.DockerVersion]sdkclient.Client
+}
+
+// newVersionedClient is a variable such that the implementation can be
+// swapped out for unit tests
+var newVersionedClient = func(endpoint, version string) (sdkclient.Client, error) {
+	return docker.NewClientWithOpts(docker.WithVersion(version),docker.WithHost(endpoint))
+}
+
+// NewFactory initializes a client factory using a specified endpoint.
+func NewFactory(ctx context.Context, endpoint string) Factory {
+	return &factory{
+		endpoint: endpoint,
+		clients: findDockerVersions(ctx, endpoint),
+	}
+}
+
+func (f *factory) GetDefaultClient() (sdkclient.Client, error) {
+	return f.GetClient(getDefaultVersion())
+}
+
+func (f *factory) FindSupportedAPIVersions() []dockerclient.DockerVersion {
+	var supportedVersions []dockerclient.DockerVersion
+	for _, testVersion := range getAgentVersions() {
+		_, err := f.GetClient(testVersion)
+		if err != nil {
+			continue
+		}
+		supportedVersions = append(supportedVersions, testVersion)
+	}
+	return supportedVersions
+}
+
+func (f *factory) FindKnownAPIVersions() []dockerclient.DockerVersion {
+	var knownVersions []dockerclient.DockerVersion
+	for _, testVersion := range dockerclient.GetKnownAPIVersions() {
+		_, err := f.GetClient(testVersion)
+		if err != nil {
+			continue
+		}
+		knownVersions = append(knownVersions, testVersion)
+	}
+	return knownVersions
+}
+
+// FindClientAPIVersion returns the version of the client from the map
+func (f *factory) FindClientAPIVersion(client sdkclient.Client) dockerclient.DockerVersion {
```

Comment:
I think you can get rid of this by using the [ClientVersion](https://godoc.org/github.com/docker/docker/client#Client.ClientVersion) API.

Reply:
Done -- Used ClientVersion() and added new tests with mocks.